crates/ruma-signatures/src/error.rs

```diff
@@ -0,0 +1,276 @@
+use ruma_identifiers::{EventId, RoomVersionId};
+use thiserror::Error;
+
+/// `ruma-signature`'s error type, wraps a number of other error types.
+#[derive(Debug, Error)]
+#[non_exhaustive]
+pub enum Error {
+    /// [`JsonError`] wrapper.
+    #[error("JSON error: {0}")]
+    Json(#[from] JsonError),
+
+    /// [`VerificationError`] wrapper.
+    #[error("Verification error: {0}")]
+    Verification(#[from] VerificationError),
+
+    /// [`ParseError`] wrapper.
+    #[error("Parse error: {0}")]
+    Parse(#[from] ParseError),
+
+    /// Wrapper for [`pkcs8::der::Error`].
+    #[error("DER Parse error: {0}")]
+    DerParse(pkcs8::der::Error),
+
+    /// [`SplitError`] wrapper.
+    #[error("Split error: {0}")]
+    SplitError(#[from] SplitError),
+}
+
+/// All errors related to JSON validation/parsing.
+#[derive(Debug, Error)]
+#[non_exhaustive]
+pub enum JsonError {
+    /// Signals that `target` is not of type `of_type` ([`JsonType`]).
+    #[error("Value in {target:?} must be a JSON {of_type:?}")]
+    NotOfType {
+        /// An arbitrary "target" that doesn't have the required type.
+        target: String,
+        /// The JSON type the target was expected to be.
+        of_type: JsonType,
+    },
+
+    /// Like [`JsonError::NotOfType`], only called when the `target` is a multiple;
+    /// array, set, etc.
+    #[error("Values in {target:?} must be JSON {of_type:?}s")]
+    NotMultiplesOfType {
+        /// An arbitrary "target" where
+        /// each or one of it's elements doesn't have the required type.
+        target: String,
+        /// The JSON type the element was expected to be.
+        of_type: JsonType,
+    },
+
+    /// Signals that a specific field is missing from a JSON object.
+    #[error("JSON object must contain the field {0:?}")]
+    JsonFieldMissingFromObject(String),
+
+    /// Signals a key missing from a JSON object.
+    ///
+    /// Note that this is different from [`JsonError::JsonFieldMissingFromObject`],
+    /// this error talks about an expected identifying key (`"ed25519:abcd"`)
+    /// missing from a target, where the key has a specific "type"/name.
+    #[error("JSON object {for_target:?} does not have {type_of} key {with_key:?}")]
+    JsonKeyMissing {
+        /// The target from which the key is missing.
+        for_target: String,
+        /// The kind of thing the key indicates.
+        type_of: String,
+        /// The key that is missing.
+        with_key: String,
+    },
+
+    /// A derivative error from [`ruma_serde::CanonicalJsonError`],
+    /// captured here.
+    #[error("Canonical JSON error: {0}")]
+    CanonicalJson(#[from] ruma_serde::CanonicalJsonError),
+
+    /// A more generic JSON error from [`serde_json`].
+    #[error(transparent)]
+    Serde(#[from] serde_json::Error),
+}
+
+// TODO: make macro for this
```

Comment:
I think is good as it is right now. Why do you want to create a macro for this?

Reply:
Because it's boilerplate, and it's easy to make a mistake by using the wrong struct or name for the function.

Also, with a macro it's cleaner and easier to read, IMO